### PR TITLE
Implement clustered DNS for FAN networking

### PR DESCRIPTION
**Pull request by stgraber**

### CODE REVIEW

File: lxd/networks.go [hunk 1]
```diff
@@ -1898,3 +1930,79 @@ func (n *network) Update(newNetwork api.NetworkPut) error {
...
+		}
+
+		state, err := client.GetNetworkState(n.name)
+		if err != nil {
+			time.Sleep(30 * time.Second)
+			goto again
+		}
+
+		for _, addr := range state.Addresses {
+			if addr.Family != "inet" || addr.Scope != "global" {
```
Comment (brauner) [inline lxd/networks.go]: What about `inet6`?

File: lxd/networks.go [hunk 2]
```diff
@@ -1898,3 +1930,79 @@ func (n *network) Update(newNetwork api.NetworkPut) error {
...
+
+	err = cmd.Start()
+	if err != nil {
+		logger.Errorf("Failed to start forkdns for network '%s': %v", n.name, err)
+		return err
+	}
+
+	// Write the PID file
+	pidPath := shared.VarPath("networks", n.name, "forkdns.pid")
+	err = ioutil.WriteFile(pidPath, []byte(fmt.Sprintf("%d\n", cmd.Process.Pid)), 0600)
```
Comment (brauner) [inline lxd/networks.go]: If this fails you don't really have a way of killing the daemon later. Shouldn't you kill the  `cmd.Start()` process or are you doing it somewhere else?

File: lxd/networks.go (hunk 1 above)
Comment (stgraber) [reply lxd/networks.go]: Fan bridges are guaranteed IPv4-only by design.

File: lxd/networks_utils.go [hunk 3]
```diff
@@ -659,6 +660,63 @@ func networkFanAddress(underlay *net.IPNet, overlay *net.IPNet) (string, string,
...
+	pid := strings.TrimSpace(string(content))
+
+	// Check for empty string
+	if pid == "" {
+		os.Remove(pidPath)
+		return nil
+	}
+
+	// Check if the process still exists
+	if !shared.PathExists(fmt.Sprintf("/proc/%s", pid)) {
```
Comment (brauner) [inline lxd/networks_utils.go]: He, that's racy. The `PID` could've been recycled. :)

File: lxd/networks_utils.go [hunk 4]
```diff
@@ -659,6 +660,63 @@ func networkFanAddress(underlay *net.IPNet, overlay *net.IPNet) (string, string,
...
+		os.Remove(pidPath)
+		return nil
+	}
+
+	// Check if it's dnsmasq
+	cmdArgs, err := ioutil.ReadFile(fmt.Sprintf("/proc/%s/cmdline", pid))
```
Comment (brauner) [inline lxd/networks_utils.go]: Yeah, I think you should drop the check before and only perform this check. :)

File: lxd/networks.go (hunk 2 above)
Comment (stgraber) [reply lxd/networks.go]: fixed

File: lxd/networks_utils.go (hunk 4 above)
Comment (stgraber) [reply lxd/networks_utils.go]: done

**Pull request merged**